pkg/util/csiconfig.go

```diff
@@ -0,0 +1,97 @@
+/*
+Copyright 2019 The Ceph-CSI Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package util
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"k8s.io/klog"
+	"strings"
+)
+
+/*
+Mons returns a comma separated MON list from the csi config for the given clusterID
+Expected JSON structure in the passed in config file is,
+{
+	"<cluster-id>":
+	{
+		"monitors":
+		[
+			"<monitor-value>",
+			"<monitor-value>",
+			...
+		]
+	},
+	...
+}
+NOTE: JSON parsing is not done using strongly typed JSON structure, as key names are variables.
+*/
+func Mons(pathToConfig, clusterID string) (string, error) {
+	var csiConfig map[string]interface{}
+
+	// #nosec
+	content, err := ioutil.ReadFile(pathToConfig)
+	if err != nil {
+		err = fmt.Errorf("error fetching configuration for cluster ID (%s). (%s)", clusterID, err)
+		return "", err
+	}
+
+	err = json.Unmarshal(content, &csiConfig)
+	if err != nil {
+		klog.Errorf("failed to parse JSON output of csi configuration (%+v): (%+v)",
+			err, content)
+		return "", fmt.Errorf("unmarshal failed: %+v. raw buffer response: %s",
+			err, string(content))
+	}
+
+	for key, value := range csiConfig {
+		if key == clusterID {
+			switch value := value.(type) {
+			case map[string]interface{}:
+				for ckey, cvalue := range value {
+					if ckey == "monitors" {
+						switch cvalue := cvalue.(type) {
+						case []interface{}:
+							var mvalue []string
+							for _, mon := range cvalue {
+								switch mon := mon.(type) {
+								case string:
+									mvalue = append(mvalue, mon)
+								default:
+									return "", fmt.Errorf("malformed clusterID section in csi"+
+										" configuration (%v)", mvalue)
+								}
+							}
+							return strings.Join(mvalue, ","), nil
+						default:
+							return "", fmt.Errorf("malformed clusterID section in csi"+
+								" configuration (%v)", cvalue)
+						}
+					}
```

Comment:
This block has to be simplified, series of `for, if, switch` does not looks good from coding pov.

Reply:
This block should at least have had better comments, anyway here is the problem and possible solutions,

The JSON structure that I have chosen has variable key names, as the key names are the `<clusterID>`, which makes it difficult to parse into a strongly typed structure in go, and hence has to be parsed as an interface and then type checked and parsed further, leading to the loops above.

The solution to simplifying the code here would be as follows,
- Write a series of nested functions, each dealing with a part of the type casting and passing it further
OR,
- Is there a way on go to parse such structures more elegantly?
OR,
- Change the config JSON to the following known key values, and use a strongly typed JSON structure to parse the same,
```
[
    {
      "clusterID": "<clusterID>",
      "monitors": [
          "<MONValue1>",
          "<MONValue2>",
          ...
          "<MONValueN>"
      ]
    },
    ...
]
```

I prefer changing the JSON to a strongly types structure, keeps it simple for other config managers to parse and update as well.

Let me know if you have a preference or another solution for the problem.

Reply:
Changed configuration to reflect new JSON structure instead, thereby simplifying the code here.